src/os/bluestore/BlueStore.cc

```diff
@@ -596,6 +617,121 @@ ostream& operator<<(ostream& out, const BlueStore::Buffer& b)
   return out << ")";
 }
 
+namespace {
+
+/*
+ * Due to a bug in key string encoding (see a comment for append_escaped)
+ * the KeyValueDB iterator does not lexicographically sort the same
+ * way that ghobject_t does: objects with the same hash may have wrong order.
+ *
+ * This is the iterator wrapper that fixes the keys order.
+ */
+
+class CollectionListIterator {
+public:
+  CollectionListIterator(const KeyValueDB::Iterator &it)
+    : m_it(it), m_chunk_iter(m_chunk.begin()) {
+  }
+
+  bool valid() const {
+    return m_key != nullptr;
+  }
+
+  const ghobject_t &oid() const {
+    return *m_oid;
+  }
+
+  const std::string &key() const {
+    return *m_key;
+  }
+
+  void lower_bound(const std::string &key) {
+    ghobject_t oid;
+    int r = get_key_object(key, &oid);
+    ceph_assert(r != -1);
+
+    std::string pkey;
+    _key_encode_prefix(oid, &pkey);
+
+    m_it->lower_bound(pkey);
+    if (!get_next_chunk()) {
+      return;
+    }
+
+    auto iter = m_chunk.find(oid);
+    if (iter != m_chunk.end()) {
+      m_chunk_iter = iter;
+    }
+
+    m_oid = &m_chunk_iter->first;
+    m_key = &m_chunk_iter->second;
+    m_chunk_iter++;
+  }
+
+  void upper_bound(const std::string &key) {
+    lower_bound(key);
+
+    if (valid() && *m_key == key) {
+      next();
+    }
+  }
+
+  void next() {
+    if (m_chunk_iter == m_chunk.end()) {
+      if (!get_next_chunk()) {
```

Comment:
isn't get_next_chunk() call redundant? get_next_chunk() func looks like a single-shot stuff. It wouldn't retrieve additional data after the first call unless KV's iterator is reset by lower_bound. Hence no need for its call from next() func.

Reply:
I am not sure I understand your point. A user (bluestore) may call `upper/lower_bound()` and reset the KV iterator), and may just call `next()` and then if the chunk is fully processed we need to load the next chunk (advancing the KV iterator, which `get_next_chunk()` does, see `m_it->next()` there).

Reply:
never mind, I missed that  loop get_next_chunk's loop terminates when hash value changes...

Reply:
Indeed. Thanks. Updated.